### PR TITLE
restAPI scaffolding

### PR DESCRIPTION
**Pull request by Unicus-Tom**
Scaffolding completed with 3 tests.
Resolves #28

### CODE REVIEW

File: restapi.test/restapi.test.csproj [hunk 1]
```diff
@@ -0,0 +1,20 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (gjersvik) [inline restapi.test/restapi.test.csproj]: Filename should have been "Sepes.RestApi.Tests.csproj" And the root namespace should be the same.
The directory should be called "Sepes.RestApi.Tests" so full path: restapi/Sepes.Restapi.Tests/Sepes.RestApi.Tests.csproj

File: restapi.test/restapi_AlwaysPasses.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
+using NUnit.Framework;
```
Comment (gjersvik) [inline restapi.test/restapi_AlwaysPasses.cs]: Just remove this file it serves no purpose.

File: restapi.test/restapi_AlwaysFails.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
+using NUnit.Framework;
```
Comment (gjersvik) [inline restapi.test/restapi_AlwaysFails.cs]: This test must be removed the first rule of automated testing: No failing test is allowed. If failing tests becomes the norm people will just start to ignore the test results.

File: restapi.test/restapi_SimpleUnitTest.cs [hunk 4]
```diff
@@ -0,0 +1,30 @@
+using Microsoft.AspNetCore.Mvc;
+using NUnit.Framework;
+using Sepes.restapi.ValuesController; //Add the controller to this test
+using System.Linq; //Adds some mathematics
+
+
+namespace Tests
+{
+
+    public class SimpleUnitTest
```
Comment (gjersvik) [inline restapi.test/restapi_SimpleUnitTest.cs]: Tests should have the name of the class under test with as Tests appended. Should be ValuesControllerTests

File: restapi.test/restapi_SimpleUnitTest.cs [hunk 5]
```diff
@@ -0,0 +1,30 @@
+using Microsoft.AspNetCore.Mvc;
+using NUnit.Framework;
+using Sepes.restapi.ValuesController; //Add the controller to this test
+using System.Linq; //Adds some mathematics
+
+
+namespace Tests
```
Comment (gjersvik) [inline restapi.test/restapi_SimpleUnitTest.cs]: Namespace should mirror project+path so Sepes.Restapi.Tests.Controllers

File: restapi.test/restapi_SimpleUnitTest.cs [hunk 6]
```diff
@@ -0,0 +1,30 @@
+using Microsoft.AspNetCore.Mvc;
```
Comment (gjersvik) [inline restapi.test/restapi_SimpleUnitTest.cs]: The directory structure of Unit Tests should mirror that of the project under tests. So this file should be (fullpath):
restapi/Sepes.RestApi.Tests/Controllers/ValuesControllerTest.cs

File: restapi.test/restapi_SimpleUnitTest.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+using NUnit.Framework;
+using Sepes.restapi.ValuesController; //Add the controller to this test
```
Comment (gjersvik) [inline restapi.test/restapi_SimpleUnitTest.cs]: This one will need to change when you fix other issues later. To using Sepes.RestApi.Controllers;

File: restapi.test/restapi_SimpleUnitTest.cs [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+using System.Linq; //Adds some mathematics
+
+
+namespace Tests
+{
+
+    public class SimpleUnitTest
+    {
+        [SetUp]
```
Comment (gjersvik) [inline restapi.test/restapi_SimpleUnitTest.cs]: Just remove empty setup method serves no purpose.

File: restapi.test/restapi_SimpleUnitTest.cs [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+        public void Setup()
+        {
+        }
+
+        [Test]
+        public void AddNumbers()
+        {
+            int[] testProducts = {2,3};
+            var controller = new ValuesController(); //Make the controller object
+            ActionResult<int> result = controller.AddNumbers(testProducts[0], testProducts[1]); //Call funtion and pass in variables
```
Comment (gjersvik) [inline restapi.test/restapi_SimpleUnitTest.cs]: can just use var here. Also no need for a testProducts array you can just pass the values in.
Should be: var result = controller.AddNumbers(2, 3);

File: restapi.test/restapi_SimpleUnitTest.cs [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+            int resultReal = testProducts.Sum();          
+
+
+            Assert.AreEqual(resultReal, result.Value);
```
Comment (gjersvik) [inline restapi.test/restapi_SimpleUnitTest.cs]: Its fine to use hard coded values in test. And the value under test should be first.
Assert.AreEqual(result.Value, 5);

With that you can remove resultReal,  testProducts and using System.Linq;

File: restapi.test/restapi.test.csproj [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+
+  <ItemGroup>
+    <PackageReference Include="microsoft.aspnetcore.app" Version="2.2.5" />
+    <PackageReference Include="nunit" Version="3.12.0" />
+    <PackageReference Include="NUnit3TestAdapter" Version="3.11.0" />
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="15.9.0" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\restapi\restapi.csproj" />
```
Comment (gjersvik) [inline restapi.test/restapi.test.csproj]: This line will need to be fixed when refactor the project structure.

File: restapi.test/.gitignore [hunk 12]
```diff
@@ -0,0 +1,327 @@
+
```
Comment (gjersvik) [inline restapi.test/.gitignore]: This file can be deleted as it ignore will be coverts by a common ignore for project and tests.

File: restapi/Controllers/ValuesController.cs [hunk 13]
```diff
@@ -0,0 +1,52 @@
+﻿using System;
```
Comment (gjersvik) [inline restapi/Controllers/ValuesController.cs]: Should be moved to restapi/Sepes.RestApi/Controllers/ValuesController.cs

File: restapi/Controllers/ValuesController.cs [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc;
+
+namespace Sepes.restapi.ValuesController
```
Comment (gjersvik) [inline restapi/Controllers/ValuesController.cs]: Sepes.RestApi.Controllers

File: restapi/Dockerfile [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+EXPOSE 5001
+
+# Copy everything else and build
+COPY . ./
+RUN dotnet publish -c Release -o out
+
+# Build runtime image
+FROM mcr.microsoft.com/dotnet/core/aspnet:2.2
+WORKDIR /app
+COPY --from=build-env /app/out .
```
Comment (gjersvik) [inline restapi/Dockerfile]: The EXPOSE should have been here somewhere after this last from.

File: restapi/Program.cs [hunk 16]
```diff
@@ -0,0 +1,24 @@
+﻿using System;
```
Comment (gjersvik) [inline restapi/Program.cs]: Should be moved to restapi/Sepes.RestApi/Program.cs

File: restapi/Program.cs [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+
+namespace restapi
```
Comment (gjersvik) [inline restapi/Program.cs]: Sepes.RestApi

File: restapi/Properties/launchSettings.json [hunk 18]
```diff
@@ -0,0 +1,30 @@
+﻿{
```
Comment (gjersvik) [inline restapi/Properties/launchSettings.json]: Move to restapi/Sepes.RestApi/Properties/launchSettings.json

File: restapi/Startup.cs [hunk 19]
```diff
@@ -0,0 +1,48 @@
+﻿using System;
```
Comment (gjersvik) [inline restapi/Startup.cs]: Move to restapi/Sepes.RestApi/Startup.cs

File: restapi/Startup.cs [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.HttpsPolicy;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace restapi
```
Comment (gjersvik) [inline restapi/Startup.cs]: Sepes.RestApi

File: restapi/TODO.txt [hunk 21]
```diff
@@ -0,0 +1,4 @@
+
```
Comment (gjersvik) [inline restapi/TODO.txt]: This should have been a guthub issue.

File: restapi/appsettings.Development.json [hunk 22]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (gjersvik) [inline restapi/appsettings.Development.json]: Move to restapi/Sepes.RestApi/appsettings.Development.json

File: restapi/appsettings.json [hunk 23]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (gjersvik) [inline restapi/appsettings.json]: Move to restapi/Sepes.RestApi/appsettings.json

File: restapi/restapi.csproj [hunk 24]
```diff
@@ -0,0 +1,14 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
```
Comment (gjersvik) [inline restapi/restapi.csproj]: Move/rename to restapi/Sepes.RestApi/Sepes.RestApi.csproj

File: restapi/restapi.csproj [hunk 25]
```diff
@@ -0,0 +1,14 @@
...
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp2.2</TargetFramework>
+    <AspNetCoreHostingModel>InProcess</AspNetCoreHostingModel>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.AspNetCore.App" />
+    <PackageReference Include="Microsoft.AspNetCore.Razor.Design" Version="2.2.0" PrivateAssets="All" />
+    <PackageReference Include="NUnit" Version="3.12.0" />
```
Comment (gjersvik) [inline restapi/restapi.csproj]: Remove this dependency its needed. We have a test project now.

File: sepes.sln [hunk 26]
```diff
@@ -0,0 +1,48 @@
+﻿
```
Comment (gjersvik) [inline sepes.sln]: This file should be called Sepes.RestApi.sln.
The changes we have made will totally invalidate this file.
Do not try to edit use tools to fix or generate new one from scratch.

File: restapi/Dockerfile [hunk 27]
```diff
@@ -0,0 +1,19 @@
+FROM mcr.microsoft.com/dotnet/core/sdk:2.2 AS build-env
```
Comment (gjersvik) [inline restapi/Dockerfile]: This file is in the right place but will most likely stop working as we move things around.
But do not of the top of my head know how to fix it.

**Review by gjersvik — CHANGES_REQUESTED** [review]

File: restapi.test/restapi_SimpleUnitTest.cs (hunk 9 above)
Comment (Unicus-Tom) [reply restapi.test/restapi_SimpleUnitTest.cs]: Yep, array was leftover from the debugging and attempts to emulate microsoft example.
It was left inside with the idea it can be used to make test accept dynamic input for testing of edge values. But I agree its unecesarry for this as no production logic exists

**Review by gjersvik — APPROVED** [review]
Looks great nice structure.

File: restapi.test/restapi_SimpleUnitTest.cs (hunk 10 above)
Comment (gjersvik) [reply restapi.test/restapi_SimpleUnitTest.cs]: It seems i was wrong on this one.
